### PR TITLE
DO NOT MERGE: Regex source generator prototype

### PR DESCRIPTION
**Pull request by pgovind**
This is just proof of concept that duplicating the current IL in RegexCompiler.cs as C# in a StringBuilder and it to the compilation as a source generator works!

Approaches considered:
1. Duplicating IL as C#
2. Abstraction in RegexCompiler.cs to output C#/IL. For ex: `Ldloc(_runtextposLocal)` => `runtextpos`: This would be a more elegant approach, but it is not straightforward to implement. We'd have to come up with a way to locally store state to handle the branching and goto statements implemented in the IL. I'm also not certain that the investment here is worth the result for S.T.Regex. In any case, we can also implement this in the future if we really want it. 
3. IL -> C# decompiler: While true that we only use simple IL statements, this will still likely require more time investment than is worth it IMO.
4. Write IL in the source generator: Roslyn doesn't support this yet. 

I decided to go with 1 for the reasons mentioned. Discussion here: https://github.com/dotnet/runtime/issues/44676#issuecomment-761249170

Current Prototype:
Adds a new `public static void CompileWithSourceGenerator(RegexCompilationInfo)` (will be `internal` in production) method on `Regex`. `RegexCompilationInfo` itself has 3 member `StringBuilder` fields to output the `DerivedRegex`, `DerivedRegexFactory` and `DerivedRegexRunner` classes. The source generator just calls `CompileWithSourceGenerator` with the given pattern and adds the 3 `StringBuilders` to the compilation. It lives here: https://github.com/pgovind/RegexSourceGenerator. For the prototype, the `pattern` is hard-coded in the source generator. For the product code, we will likely need a new `Attribute` that the source generator looks for in its `Initialize` method or a new enum in `RegexOptions`. Also, `FindFirstChar` and `Go` just return at the moment. I have local changes to finish transpiling `FindFirstChar` already, but it's not in this draft.

A question to consider here is where the source generator lib should live? We don't ship any source generators from dotnet/runtime, so I'm not sure the infra is set up for this yet. IIRC, @ericstj looked into this for the JSON source generator?

Notes for myself when I get back to this later:
1. VS doesn't work with SG properly yet. VS will report many errors, but the command line will build fine.  
2. `dotnet clean/build` on a console app will invoke the source generator.  Can be debugged with `Debugger.Launch()`.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexAssemblyCompiler.cs [hunk 1]
```diff
@@ -54,54 +59,107 @@ internal void GenerateRegexType(string pattern, RegexOptions options, string nam
...
-            _ilg = DefineMethod(regexRunnerTypeBuilder, "Go", null);
-            GenerateGo();
-            _ilg = DefineMethod(regexRunnerTypeBuilder, "FindFirstChar", typeof(bool));
-            GenerateFindFirstChar();
-            _ilg = DefineMethod(regexRunnerTypeBuilder, "InitTrackCount", null);
-            GenerateInitTrackCount();
-            Type runnerType = regexRunnerTypeBuilder.CreateType()!;
+            TypeBuilder regexRunnerTypeBuilder = null!;
+            Type runnerType = default;// = regexRunnerTypeBuilder.CreateType()!;
+            if (regexInfo.regexRunnerCode == null)
```
Comment (pgovind) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexAssemblyCompiler.cs:64]: Thee if-else here can likely be simplified

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler_WrongTranspile.cs [hunk 2]
```diff
@@ -0,0 +1,5778 @@
+//// Licensed to the .NET Foundation under one or more agreements.
```
Comment (pgovind) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler_WrongTranspile.cs:1]: Contains local changes to RegexCompiler.cs. This is WIP and won't compile yet. Included just for reference

**Comment by ericstj** [comment]
The prototype for JSON can be seen here https://github.com/dotnet/runtimelab/tree/feature/JsonCodeGen/src/libraries/System.Text.Json
cc @layomia

**Comment by ericstj** [comment]
It looks to me like your source generator calls in to the RegEx type in the framework that the compiler is running on.  This is usually a bad thing since build framework does not always equal TargetFramework. Did you consider carrying a subset of the RegEx code in the source generator so that it wouldn’t need to call into the framework?

**Comment by pgovind** [comment]
> This is usually a bad thing since build framework does not always equal TargetFramework
 
 What exactly does this mean? What is build framework?

**Comment by ericstj** [comment]
The framework that CSC executes on is the build framework.  The framework the application/ library targets is the TargetFramework.

**Comment by danmosemsft** [comment]
The interface between engine and generated regex is stable (so far) so in theory the generator can generate a regex that works on any runtime Framework forwards and backwards. However it would be easier to reason about and test the SG if it minimized its dependency on the SDK implementation. That might also allow it to be updated out of band more easily.

One thing to consider is if we want to extend that interface in future: the SG will need to consult the references going to the compiler in order to know whether it has the new interfaces available.

**Comment by stephentoub** [comment]
> Did you consider carrying a subset of the RegEx code in the source generator so that it wouldn’t need to call into the framework?

This is the path I thought we'd be on.  We'd effectively build the parser / node model code into both S.T.RegularExpressions.dll and S.T.RegularExpressions.Generator.dll, and then the processing of the output of the parser would live entirely in S.T.R.Generator.dll.  Assuming we're on the "just output C# directly path", nothing about the structure of the RegexCompiler / RegexAssemblyCompiler / etc. would need to be maintained or modified; we'd just have a new RegexSourceCompiler type in the new assembly that would be entirely responsible for translating parser output => C# source string.  If we change gears and go down a "share abstractions for generating, where one concrete implementation spits out IL and another spits out C#", then we'd have to refactor and share more.

**Comment by pgovind** [comment]
> bad thing since build framework does not always equal TargetFramework
> However it would be easier to reason about and test the SG if it minimized its dependency on the SDK implementation
> We'd effectively build the parser / node model code into both S.T.RegularExpressions.dll and S.T.RegularExpressions.Generator.dll

Ah I see, I was still getting a feel for how a RegEx SG might work here, so I didn't fully consider this. I agree, I'll remove any dependencies on the sdk implementation other than the existing public APIs. Making these changes is straightforward. I just need a few wrappers and need to move the changes to `RegexCompiler.cs` to a new method in the source generator.

**Comment by layomia** [comment]
> A question to consider here is where the source generator lib should live? We don't ship any source generators from dotnet/runtime, so I'm not sure the infra is set up for this yet. IIRC, @ericstj looked into this for the JSON source generator?

> The prototype for JSON can be seen here https://github.com/dotnet/runtimelab/tree/feature/JsonCodeGen/src/libraries/System.Text.Json
> cc @layomia

I plan to tackle ingesting the JSON source generator into dotnet/runtime after it's design and API changes have been reviewed (and approved). The two considerations, as I see it, are 

- layout (where the code lives)
- how the generator is shipped in the product.

The proposed layout for JSON is shown in the link from Eric above.

For shipping - the prototype for the JSON source generator produces an [OOB package](https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet-experimental&view=versions&package=System.Text.Json&version=6.0.0-s390x.1.20553.1&protocolType=NuGet) that contains the updated `System.Text.Json.dll` with public API changes needed by the generator, and also the `System.Text.Json.SourceGeneration.dll` analyzer/generator. This is helpful to get feedback from the community/partners as the effort goes on. When the generator is ingested into dotnet/runtime, we'll maintain this shipping vehicle for OOB scenarios. For .NET 6 (in-box scenarios) we'll want the generator to flow into the SDK. I haven't looked into this yet. cc @ericstj, @joperezr 

There's still a good amount of [work](https://github.com/dotnet/runtimelab/projects/1) to be done on the JSON source generator over the next several weeks. I imagine other generators, like Regex, might have established precedent for ingesting source generators into dotnet/runtime by the time I circle back to this.

**Comment by jeffhandley** [comment]
Here's the issue that will track the implementation: https://github.com/dotnet/runtime/issues/44676

**Pull request closed without merging**